ospf6d/ospf6_bfd.c

```diff
@@ -118,134 +87,37 @@ void ospf6_bfd_trigger_event(struct ospf6_neighbor *on, int old_state,
  *                               zebra for starting/stopping the monitoring of
  *                               the neighbor rechahability.
  */
-static void ospf6_bfd_reg_dereg_all_nbr(struct ospf6_interface *oi, int command)
+static void ospf6_bfd_reg_dereg_all_nbr(struct ospf6_interface *oi, bool enable)
 {
 	struct ospf6_neighbor *on;
 	struct listnode *node;
 
 	for (ALL_LIST_ELEMENTS_RO(oi->neighbor_list, node, on)) {
-		if (command == ZEBRA_BFD_DEST_REGISTER)
-			ospf6_bfd_info_nbr_create(oi, on);
-
+		/*
+		 * If not connected yet, don't create any session but defer it
+		 * for later.
+		 */
 		if (on->state < OSPF6_NEIGHBOR_TWOWAY) {
-			if (command == ZEBRA_BFD_DEST_DEREGISTER)
-				bfd_info_free(
-					(struct bfd_info **)&on->bfd_info);
+			bfd_sess_enable(on->bfd_session, false);
 			continue;
 		}
 
-		ospf6_bfd_reg_dereg_nbr(on, command);
-	}
-}
-
-/*
- * ospf6_bfd_nbr_replay - Replay all the neighbors that have BFD enabled
- *                        to zebra
- */
-static int ospf6_bfd_nbr_replay(ZAPI_CALLBACK_ARGS)
-{
-	struct vrf *vrf = vrf_lookup_by_id(vrf_id);
-	struct listnode *node;
-	struct interface *ifp;
-	struct ospf6_interface *oi;
-	struct ospf6_neighbor *on;
-	char dst[64];
-
-	if (IS_OSPF6_DEBUG_ZEBRA(RECV))
-		zlog_debug("Zebra: BFD Dest replay request");
-
-	/* Send the client registration */
-	bfd_client_sendmsg(zclient, ZEBRA_BFD_CLIENT_REGISTER, vrf_id);
-
-	/* Replay the neighbor, if BFD is enabled on the interface*/
-	FOR_ALL_INTERFACES (vrf, ifp) {
-		oi = (struct ospf6_interface *)ifp->info;
-
-		if (!oi || !oi->bfd_info)
-			continue;
-
-		for (ALL_LIST_ELEMENTS_RO(oi->neighbor_list, node, on)) {
-			if (on->state < OSPF6_NEIGHBOR_TWOWAY)
-				continue;
-
-			if (IS_OSPF6_DEBUG_ZEBRA(SEND)) {
-				inet_ntop(AF_INET6, &on->linklocal_addr, dst,
-					  sizeof(dst));
-				zlog_debug("Replaying nbr (%s) to BFD", dst);
-			}
-
-			ospf6_bfd_reg_dereg_nbr(on, ZEBRA_BFD_DEST_UPDATE);
-		}
+		ospf6_bfd_info_nbr_create(oi, on);
+		bfd_sess_enable(on->bfd_session, enable);
+		bfd_sess_install(on->bfd_session);
```

Comment:
This is for the "enable" case. The "disable" case is not processed. We should free the BFD session in this case.

Reply:
This is fine... this is a generalization to avoid code duplication.

If `enable = false` then install will do nothing:

```c
void bfd_sess_install(struct bfd_session_params *bsp)
{
	/* Don't attempt to install/update when disabled. */
	if (!bsp->enabled)
		return;

	bsp->lastev = BSE_INSTALL;
	thread_add_event(bsglobal.tm, _bfd_sess_send, bsp, 0, &bsp->installev);
}
```

Reply:
There should be code to free the BFD session somewhere. Currently the session is never freed.

Reply:
The session is `free()`d when:

1. Session is removed ( https://github.com/opensourcerouting/frr/blob/ospfv3-bfd-rework/ospf6d/ospf6_neighbor.c#L153 )
2. Interface is configuration is removed ( https://github.com/opensourcerouting/frr/blob/ospfv3-bfd-rework/ospf6d/ospf6_interface.c#L246 )
3. On shutdown when interface configurations are purged ( https://github.com/opensourcerouting/frr/blob/ospfv3-bfd-rework/ospf6d/ospf6_main.c#L96 )

I think you mean when the BFD configuration is removed... In this case I need to think how to `free` the memory without just deactivating it. Or did I miss somewhere else?

Reply:
Sure, sorry, I mean when the config is removed. We do this in other daemons.
I think just adding a separate case to `ospf6_bfd_reg_dereg_all_nbr` should be enough:
```
		if (enable) {
			ospf6_bfd_info_nbr_create(oi, on);
			bfd_sess_enable(on->bfd_session, enable);
			bfd_sess_install(on->bfd_session);
		} else {
			bfd_sess_free(&on->bfd_session);
		}
```

Reply:
Fixed